cdap-security/src/main/java/co/cask/cdap/security/store/secretmanager/SecretManagerSecureStoreService.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.security.store.secretmanager;
+
+import co.cask.cdap.api.security.store.SecureStoreData;
+import co.cask.cdap.api.security.store.SecureStoreMetadata;
+import co.cask.cdap.common.NamespaceNotFoundException;
+import co.cask.cdap.common.SecureKeyNotFoundException;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.namespace.NamespaceQueryAdmin;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.SecureKeyId;
+import co.cask.cdap.securestore.spi.SecretManager;
+import co.cask.cdap.securestore.spi.SecretManagerContext;
+import co.cask.cdap.securestore.spi.SecretNotFoundException;
+import co.cask.cdap.securestore.spi.secret.Secret;
+import co.cask.cdap.securestore.spi.secret.SecretMetadata;
+import co.cask.cdap.security.store.SecureStoreService;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Secret manager secure store service to initialize {@link SecretManager}.
+ */
+public class SecretManagerSecureStoreService extends AbstractIdleService implements SecureStoreService {
+  private static final Logger LOG = LoggerFactory.getLogger(SecretManagerSecureStoreService.class);
+
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+  private final SecretManagerContext context;
+  private final String type;
+  private SecretManager secretManager;
+
+  @Inject
+  public SecretManagerSecureStoreService(CConfiguration cConf, NamespaceQueryAdmin namespaceQueryAdmin) {
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
```

Comment:
It is better to call another constructor instead.

Reply:
This is because we create objects in this constructor, otherwise it makes it harder to read.

Reply:
? You mean this is hard to read?

```java
this(
  namespaceQueryAdmin,
  new DefaultSecretManagerContext(cConf),
  cConf.get(Constants.Security.Store.PROVIDER),
  new SecretManagerExtensionLoader(cConf.get(Constants.Security.Store.EXTENSIONS_DIR)).get(type)
);
```